[RLlib; docs] New API stack migration guide. (#47779)
ray-project/ray: doc/source/rllib/new-api-stack-migration-guide.rst

.. _new-api-stack-migration-guide:

New API Stack Migration Guide
=============================

This page explains - step by step - how to convert and translate your existing old API stack
RLlib classes and code to RLlib's new API stack.
:ref:`Read here on what the new API stack is and why you should migrate soon<rllib-new-api-stack-guide>`.

.. note::

    Even though the new API stack still rudimentary supports `TensorFlow <https://tensorflow.org>`__ and
    has been written in a framework-agnostic fashion, RLlib will soon move to `PyTorch <https://pytorch.org>`__
    as the only supported deep learning framework (dropping TensorFlow support entirely).

Change your AlgorithmConfig
---------------------------

The new API stack is disabled by default for all RLlib algorithms. To activate it, use the `api_stack()` method
in your `AlgorithmConfig` object like so:

.. testcode::

    from ray.rllib.algorithms.ppo import PPOConfig

    config = (
        PPOConfig()
        # Switch both the new API stack flags to True (both False by default).
        # This enables the use of
        # a) RLModule (replaces ModelV2) and Learner (replaces Policy)

>         # b) the correct EnvRunner (single-agent vs multi-agent) and ConnectorV2 pipelines.

Author: done

and its new stack counterpart. The following sections list all important differences.
Go through the list of changes and make sure you are either translating the respective
settings or drop them altogether, if they are no longer supported or needed in the new API stack:

AlgorithmConfig.framework()
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Even though the new API stack still rudimentary supports `TensorFlow <https://tensorflow.org>`__, RLlib will soon move to
`PyTorch <https://pytorch.org>`__ as the only supported deep learning framework.

This means that the following framework-related settings will be deprecated when using the new API stack:

.. testcode::

    config.framework(
        eager_tracing=True,
        eager_max_retraces=20,
        tf_session_args={},
        local_tf_session_args={},
    )

AlgorithmConfig.resources()
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The `num_gpus` and `_fake_gpus` settings have been deprecated. In order to place your RLModule on one or more GPUs on the Learner side,
do the following:

.. testcode::

    # The following setting is equivalent to the old stack's `config.resources(num_gpus=2)`.
    config.learners(
        num_learners=2,

>         num_gpus_per_learner=1,

Author: done

:py:class:`~ray.rllib.core.learner.learner.Learner`:

.. testcode::

    config.training(
        train_batch_size_per_learner=512,
    )

This way, you won't need to touch this setting, even when increasing the number of :py:class:`~ray.rllib.core.learner.learner.Learner`
(through `config.learners(num_learners=...)`).

Neural Network Config
.....................

The old stack's `config.training(model=...)` is no longer supported on the new stack.
Instead, use the new stack's :py:meth:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig.rl_module`
method to configure an RLlib default RLModule or specify (and configure) a custom one.

:ref:`See here for more details on how to configure your RLModule(s) <rlmodule-guide>`.

Learner Config
..............

Learning Rate- and Coefficient Schedules
........................................

If you are using schedules for your learning rate or other coefficients (for example the
`entropy_coeff` setting in PPO), provide scheduling information directly in the respective setting.
There is no specific, separate setting anymore for scheduling behavior.

> When defining a schedule, provide a list of 2-tuples, where the first item is the global timestep

Author: done. And linked to example script.

Another example: To create a entropy coefficient schedule that starts with a value of 0.05, then increases over 1M timesteps to 0.1 and
then suddenly drops to 0 (after the 1Mth timestep), do:

.. testcode::

    config.training(
        entropy_coeff=[
            [0, 0.05],  # <- initial value at timestep 0
            [1000000, 0.1],  # <- value at 1M timesteps
            [1000001, 0.0],  # <- sudden drop to 0.0 right after 1M timesteps
        ]
    )

AlgorithmConfig.env_runners()
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. testcode::

    # RolloutWorkers have been re-written to EnvRunners:
    config.env_runners(
        num_env_runners=2,  # use this instead of `num_workers`
    )

    # The following `env_runners` settings are deprecated and should no longer be explicitly
    # set on the new stack:
    config.env_runners(
        create_env_on_local_worker=False,
        sample_collector=None,
        enable_connectors=True,
        remote_worker_envs=False
        remote_env_batch_wait_ms=0,
        preprocessor_pref="deepmind",
        enable_tf1_exec_eagerly=False,
        sampler_perf_stats_ema_coef=None,
    )

> In case you were using the `observation_filter` setting, perform the following translations:

Author: Doesn't exist yet :(

pipelines on the new API stack.

TODO: Continue here!!

ModelV2 to RLModule
-------------------

In case you are using a custom :py:class:`~ray.rllib.models.modelv2.ModelV2` class and would like to translate
the entire NN architecture and possibly action distribution logic to the new API stack, take a look at
the :ref:`RLModule documentation <rlmodule-guide>` first, then come back to this location here.

Here are also helpful example scripts on `how to write a custom CNN-containing RLModule <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/custom_cnn_rl_module.py>`__
and `how to write a custom LSTM-containing RLModule <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/custom_lstm_rl_module.py>`__.

Also, there are different options for translating an existing, custom :py:class:`~ray.rllib.models.modelv2.ModelV2` (old API stack)
to the new API stack's :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`:

1) You lift your ModelV2 code and drop it into a new, custom RLModule class (see the :ref:`RLModule documentation <rlmodule-guide>` for details).

> 1) You use an Algorithm checkpoint or a Policy checkpoint that you have from an old API stack training run and use this with the `new stack RLModule convenience wrapper <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/migrate_modelv2_to_new_api_stack_by_policy_checkpoint.py>`__.

Author: This is correct. Sphinx will automatically enumerate these.

Neural network configuration
............................

The old stack's `config.training(model=...)` is no longer supported on the new API stack.
Instead, use the new :py:meth:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig.rl_module`
method to configure RLlib's default :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`
or specify and configure a custom :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`.

See :ref:`RLModules API <rlmodule-guide>`, a general guide that also explains
the use of the `config.rl_module()` method.

If you have an old stack `ModelV2` and want to migrate the entire NN logic to the
new stack, see :ref:`ModelV2 to RLModule <rllib-modelv2-to-rlmodule>` for migration instructions.

Learning rate- and coefficient schedules
........................................

If you're using schedules for learning rate or other coefficients, for example, the
`entropy_coeff` setting in PPO, provide scheduling information directly in the respective setting.
Scheduling behavior doesn't require a specific, separate setting anymore.

When defining a schedule, provide a list of 2-tuples, where the first item is the global timestep

> (*num_env_steps_sampled_lifetime* in the reported metrics) and the second item is the value that should

Author: ```suggestion
(*num_env_steps_sampled_lifetime* in the reported metrics) and the second item is the value that the learning rate should reach at that timestep.
```

> be reached at that timestep. Always start the first 2-tuple with timestep 0. Note that RLlib linearly interpolates values between

Author: ```suggestion
Always start the first 2-tuple with timestep 0. Note that RLlib linearly interpolates values between
```

1) You use an Algorithm checkpoint or a Policy checkpoint that you have from an old API stack training run and use this with the `new stack RLModule convenience wrapper <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/migrate_modelv2_to_new_api_stack_by_policy_checkpoint.py>`__.
1) You have an :py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig` object from an old API stack training run and use this with the `new stack RLModule convenience wrapper <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/migrate_modelv2_to_new_api_stack_by_config.py>`__.

Custom Loss Functions and/or Policies
-------------------------------------

In case you are using one or more custom loss functions and/or custom (PyTorch) optimizers to train your models, instead of doing these
customizations inside the old stack's Policy class, you need to move these logic into the new API stack's
:py:class:`~ray.rllib.core.learner.learner.Learner` class.

:ref:`See here for more details on how to write a custom Learner <learner-guide>`.

> Here are also helpful example scripts on `how to write a simple custom loss function <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/custom_loss_fn_simple.py>`__

Reviewer: ```suggestion
The following example scripts show how to write:
- `a simple custom loss function <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/custom_loss_fn_simple.py>`__
```

> and `how to write a custom Learner with 2 optimizers and different learning rates for each <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/separate_vf_lr_and_optimizer.py>`__.

Reviewer: ```suggestion
- `a custom Learner with 2 optimizers and different learning rates for each <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/separate_vf_lr_and_optimizer.py>`__.
```

> Note that the Policy class is no longer supported in the new API stack. This class used to hold a

Reviewer: ```suggestion
Note that the new API stack doesn't support the Policy class. In the old stack, this class holds a
```

> neural network (now moved into :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`),

Reviewer: ```suggestion
neural network, which is the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` in the new API stack,
```

> a (old stack) connector (now moved into :py:class:`~ray.rllib.connector.connector_v2.ConnectorV2`),

Reviewer: ```suggestion
an old stack connector, which is the :py:class:`~ray.rllib.connector.connector_v2.ConnectorV2` in the new API stack,
```

> and one or more optimizers and losses (now moved into :py:class:`~ray.rllib.core.learner.learner.Learner`).

Reviewer: ```suggestion
and one or more optimizers and losses, which are the :py:class:`~ray.rllib.core.learner.learner.Learner` class in the new API stack.
```

> The RLModule API is much more flexible than the old stack's Policy API and

Reviewer: ```suggestion
The RL Module API is much more flexible than the old stack's Policy API and
```

> provides a cleaner separation-of-concerns experience (things related to action

Reviewer: ```suggestion
provides a cleaner separation-of-concerns experience. Things related to action
```

Reviewer: Consider using a more descriptive word in place of `things`.

> inference run on the EnvRunners, things related to updating run on the Learner workers)

Reviewer: ```suggestion
inference run on the EnvRunners, and things related to updating run on the Learner workers
```

> as well as superior scalability allowing training in a multi-GPU setup in any Ray cluster

Reviewer: ```suggestion
It also provides superior scalability, allowing training in a multi-GPU setup in any Ray cluster
```

> and multi-node + multi-GPU training on the `Anyscale <https://anyscale.com>`__ platform.

Reviewer: ```suggestion
and multi-node with multi-GPU training on the `Anyscale <https://anyscale.com>`__ platform.
```

> Custom (old-stack) Connectors

Reviewer: ```suggestion
Custom connectors (old-stack) 
```

-----------------------------

> If you are using custom (old API stack) connectors, you have to lift your logic into the

Reviewer: ```suggestion
If you're using custom connectors from the old API stack, move your logic into the
```

new :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` API.
Translate your agent connectors into env-to-module ConnectorV2 pieces and your
action connectors into module-to-env ConnectorV2 pieces.

> The :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` documentation is work in progress and linked from here shortly.

Reviewer: ```suggestion
The :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` documentation is under development.
```

> In the meantime, take a look at some examples on how to write ConnectorV2 pieces for the

Reviewer: ```suggestion
The following are some examples on how to write ConnectorV2 pieces for the
```

different pipelines:

> 1) `Example on how to perform observation frame-stacking <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/frame_stacking.py>`__.

Reviewer: ```suggestion
1) `Observation frame-stacking <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/frame_stacking.py>`__.
```

> 1) `Example on how to add the most recent action and reward to the RLModule's input <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/prev_actions_prev_rewards.py>`__.

Reviewer: ```suggestion
1) `Add the most recent action and reward to the RL Module's input <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/prev_actions_prev_rewards.py>`__.
```

> 1) `Example on how to do mean-std filtering on all observations <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/mean_std_filtering.py>`__.

Reviewer: ```suggestion
1) `Mean-std filtering on all observations <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/mean_std_filtering.py>`__.
```

> 1) `Example on how to flatten any complex observation space to a 1D space <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/flatten_observations_dict_space.py>`__.

Reviewer: ```suggestion
1) `Flatten any complex observation space to a 1D space <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/flatten_observations_dict_space.py>`__.
```
